MEN-5449: AWS IoT Core integration

tranchitella opened the pull request.

tranchitella commented:
@mender-test-bot start pipeline

tranchitella commented:
@mender-test-bot start pipeline

merlin-northern commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+
+	var deviceResp *Device
+	if err == nil {
+		deviceResp = &Device{
+			ID:          *resp.ThingId,
+			Name:        *resp.ThingName,
+			Status:      device.Status,
+			PrivateKey:  string(privateKeyToPem(privKey)),
+			Certificate: *respCert.CertificatePem,
+		}
```
similar block is repeated, maybe a cool method in the Device to "load"/"create" it from resp?
why we are not passing version here?

tranchitella replied:
Honestly, I don't think so. Not too much is shared between the two code blocks, only ID/Name.

merlin-northern commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+
+	var respListThingPrincipals *iot.ListThingPrincipalsOutput
+	if err == nil {
+		respListThingPrincipals, err = svc.ListThingPrincipalsWithContext(ctx,
+			&iot.ListThingPrincipalsInput{
+				ThingName: aws.String(deviceID),
+			})
```
nit: exact the same is repeated twice, perhaps a helper function?

tranchitella replied:
Don't know, this is going to increase the code complexity for nothing (at the end, you replace one SDK func call with a helper func call).

merlin-northern commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+				ExpectedVersion: aws.Int64(*respDescribe.Version),
+			})
+	}
+	if err != nil {
+		if _, ok := err.(*iot.ResourceNotFoundException); ok {
+			err = ErrDeviceNotFound
+		}
+		return err
+	}
```
cant we/should we bail out earlier? I mean: for instance if `DetachThingPrincipalWithContext` returns error above, there is no reason to continue, right? and we can return with error immediately (save a bit of those ifs maybe).

tranchitella replied:
Correct; on the other hand, this will lower significantly the test code coverage.
Bringing the err forward is a common pattern in golang.

merlin-northern commented on client/iotcore/client_test.go:
```diff
@@ -0,0 +1,198 @@
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/google/uuid"
+	"github.com/mendersoftware/iot-manager/model"
+	"github.com/stretchr/testify/assert"
```
nit: sorted imports look like the following:

```go
	"context"
	"flag"
	"os"
	"testing"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/google/uuid"
	"github.com/stretchr/testify/assert"

	"github.com/mendersoftware/iot-manager/model"
```

merlin-northern commented on client/iotcore/client_test.go:
```diff
@@ -0,0 +1,198 @@
+	]
+}`
+
+func init() {
```
so in order to run these tests I need to provide the AWS keys, right? CI is taking them from the gitlab settings?

tranchitella replied:
Correct; I set up the secrets in Gitlab CI/CD.

merlin-northern commented on client/iotcore/client_test.go:
```diff
@@ -0,0 +1,198 @@
+	_, err = client.GetDevice(ctx, sess, "dummy")
+	assert.EqualError(t, err, ErrDeviceNotFound.Error())
+
+	err = client.DeleteDevice(ctx, sess, device.Name)
+	assert.NoError(t, err)
```
could we have here the following scenario:

```go
	device, err := client.GetDevice(ctx, sess, deviceID)
	assert.EqualError(t, err, ErrDeviceNotFound.Error())
	assert.Nil(t, device)
```

merlin-northern commented on client/iotcore/client_test.go:
```diff
@@ -0,0 +1,198 @@
+	err = client.DeleteDevice(ctx, sess, device.Name)
+	assert.NoError(t, err)
+
+	err = client.DeleteDevice(ctx, sess, device.Name)
+	assert.EqualError(t, err, ErrDeviceNotFound.Error())
```
same as above, could we have 


```go
	device, err := client.GetDevice(ctx, sess, deviceID)
	assert.EqualError(t, err, ErrDeviceNotFound.Error())
	assert.Nil(t, device)
```

merlin-northern commented on client/iotcore/mocks/Client.go:
```diff
@@ -0,0 +1,91 @@
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+// Code generated by mockery v2.9.4. DO NOT EDIT.
```
I remember when I got instructed that we have to use one mockery version, therefore I kindly ask: do we use 2.9.4 now?

tranchitella replied:
See utils/mockgen.sh in the repo, that's where we define the version we want to use.

merlin-northern commented on client/iotcore/model.go:
```diff
@@ -0,0 +1,63 @@
+	"github.com/mendersoftware/iot-manager/model"
+)
+
+type Device struct {
+	ID            string `json:"id"`
+	Name          string `json:"name"`
+	Version       int64  `json:"version,omitempty"`
+	Status        Status `json:"status,omitempty"`
+	CertificateID string `json:"certificate_id,omitempty"`
+	Certificate   string `json:"certificate,omitempty"`
+	PrivateKey    string `json:"private_key,omitempty"`
+}
```
shouldn't we validate `Device` structure?
security sweep: how long does PrivateKey stay in memory in plain format?

tranchitella replied:
No need to validate this struct, as we never marshall/unmarshall it directly.
This is a holder of data we get/set from/to the SDK.

merlin-northern reviewed (commented):
first commit comments. I will post comments commit-by-commit if you do not mind.

merlin-northern commented on app/app_iot_core.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	if dev.Certificate != "" && dev.PrivateKey != "" {
+		err = a.wf.ProvisionExternalDevice(ctx, dev.ID, map[string]string{
+			confKeyAWSCertificate: dev.Certificate,
+			confKeyAWSPrivateKey:  dev.PrivateKey,
+		})
```
the private key will not land in the workflows logs, right? (just so I can sleep peacefully)

tranchitella replied:
Private key will land to workflows, and the get propagated to deviceconfig.
This is what we have for Azure IoT Hub as well (connection string).
We have an epic to manage secrets, but the current implementation requires this.

merlin-northern commented on app/app.go:
```diff
@@ -76,22 +79,34 @@ type App interface {
 
 // NewApp initialize a new iot-manager App
-func New(ds store.DataStore, hub iothub.Client, wf workflows.Client, da devauth.Client) App {
+func New(ds store.DataStore, wf workflows.Client, da devauth.Client) *app {
```
why are we returning struct not the interface?
because the new With* methods are not part of the App interface.
why arent they?

merlin-northern commented on app/app.go:
```diff
@@ -233,6 +248,10 @@ func (a *app) ProvisionDevice(
 			err = a.provisionIoTHubDevice(ctx, deviceID, integration)
+		case model.ProviderIoTCore:
+			err = a.provisionIoTCoreDevice(ctx, deviceID, integration, &iotcore.Device{
+				Status: iotcore.StatusDisabled,
```
being curious why do we provision in the disabled state? how do we travel to enabled state?

tranchitella replied:
Devices are always provisioned in "Disabled" state, and then get accepted when you accept the authset by the SetDeviceStatus function.

merlin-northern commented on app/app_test.go:
```diff
@@ -627,14 +483,11 @@ func TestProvisionDevice(t *testing.T) {
-			defer wf.AssertExpectations(t)
 
-			app := New(ds, hub, wf, nil)
+			app := New(ds, nil, nil)
```
why don't we need .With*(hub) here?

tranchitella replied:
No, because I moved the hub-specific tests to app_iot_hub_test.go

merlin-northern reviewed (commented):
commit 2. nice read it is.

tranchitella commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+	sess *session.Session,
+	deviceID string,
+) (*Device, error) {
+	svc := iot.New(sess)
```
We need to allocate a session each time, because each session will connect to different AWS accounts (tenant-specific).

tranchitella commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+			Status:  StatusDisabled,
+		}
+		var respListThingPrincipals *iot.ListThingPrincipalsOutput
+		respListThingPrincipals, err = svc.ListThingPrincipalsWithContext(ctx,
+			&iot.ListThingPrincipalsInput{
+				ThingName: aws.String(deviceID),
+			})
+
```
You are right, this was a mistake

tranchitella commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+			})
+
+		for _, principal := range respListThingPrincipals.Principals[:1] {
+			parts := strings.SplitAfter(*principal, "/")
```
Doesn't make any difference, I moved to "Split".

tranchitella commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+	device *Device,
+	policy string,
+) (*Device, error) {
+	svc := iot.New(sess)
+
+	awsDevice, err := c.GetDevice(ctx, sess, deviceID)
```
See reply above, we need it that way.

tranchitella commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+
+	var device *Device
+
+	if err == nil {
```
Why? This is part of the SDK contract: you either get a response, or an error.

merlin-northern replied:
since I approved we can safely argue now :)
call me an old timer, but no matter what, I believe in always checking the pointer before using it.
Re: contract: this is not java, which had contracts as it was founded on standards which you implemented. in here it is just a library that someone wrote and there is no way to enforce nor to expect anything, since it does not comply to standard you can refer to. and I know we do it all over the code base. I propose to work on a document called: _Mender coding standard_. much of the reviews ping-pong can be avoided if we had guidelines.

tranchitella commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+		cert, err := svc.DescribeCertificateWithContext(ctx, &iot.DescribeCertificateInput{
+			CertificateId: aws.String(awsDevice.CertificateID),
+		})
+		if err == nil {
```
Why? This is part of the SDK contract: you either get a response, or an error.

tranchitella commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+				ThingName: aws.String(deviceID),
+			})
+
+		for _, principal := range respListThingPrincipals.Principals[:1] {
```
If there is more than one, the device is not consistent with what we created (and we don't know how to deal with it).
I updated the code to raise an error in this case.

tranchitella commented on client/iotcore/client.go:
```diff
@@ -0,0 +1,315 @@
+	}
+
+	return pem.EncodeToMemory(csrPemBlock), nil
+}
```
Moved to crypto and added unit tests.

tranchitella commented:
@merlin-northern I addressed some of your comments here: https://github.com/mendersoftware/iot-manager/pull/63/commits/871ca4a7416156b7c36a12eac704c21cad4de142

tranchitella commented:
@mender-test-bot start pipeline --fast

merlin-northern reviewed (commented):
done with commit 99b69fd543badd0e4705be296fb28dbedc5e3ad2

merlin-northern commented on model/integration_test.go:
```diff
@@ -17,6 +17,7 @@ package model
 import (
 	"testing"
 
+	"github.com/mendersoftware/iot-manager/crypto"
 	"github.com/pkg/errors"
 	"github.com/stretchr/testify/assert"
 )
```
```go
	"github.com/pkg/errors"
	"github.com/stretchr/testify/assert"

	"github.com/mendersoftware/iot-manager/crypto"
```

merlin-northern commented on model/integration_test.go:
```diff
@@ -25,6 +26,11 @@ func str2ptr(s string) *string {
 	return &s
 }
 
+func str2cyptoptr(s string) *crypto.String {
+	c := crypto.String(s)
+	return &c
+}
+
```
do we have units for this function?

tranchitella replied:
This is a helper in unit test, it doesn't need unit tests.
It returns a pointer to crypto.String, which has its unit tests.

merlin-northern reviewed (commented):
done with commit 8b423afa99066efae69b2234c65029ed6327612a

merlin-northern commented on app/app_iot_core.go:
```diff
@@ -74,7 +81,7 @@ func (a *app) setDeviceStatusIoTCore(ctx context.Context, deviceID string, statu
 	_, err = a.iotcoreClient.UpsertDevice(ctx, sess, deviceID, &iotcore.Device{
 		Status: iotcore.NewStatusFromMenderStatus(status),
-	})
+	}, *integration.Credentials.DevicePolicyDocument)
```
nit:

```go
	_, err = a.iotcoreClient.UpsertDevice(
		ctx,
		sess,
		deviceID,
		&iotcore.Device{
			Status: iotcore.NewStatusFromMenderStatus(status),
		},
		*integration.Credentials.DevicePolicyDocument,
	)
```

merlin-northern commented on model/integration.go:
```diff
@@ -94,7 +113,8 @@ func (s Credentials) Validate() error {
 		validation.Field(&s.DevicePolicyDocument,
-			validation.When(s.Type == CredentialTypeAWS, validation.Required)),
+			validation.When(s.Type == CredentialTypeAWS, validation.Required,
+				validation.By(validatePolicy))),
```
I do not fully understand, `DevicePolicyDocument` is `*string`, so this: [return err](https://github.com/mendersoftware/iot-manager/pull/63/commits/3a46b8bba247e4b176dbd6c45a3af419c774374e#diff-8facb938c8871c2d5008b83d852f5dd59214a0032b6b76e56472ca77794af3d4R100) in `validatePolicy` is a dead code, right? is it because `validation` has to get a func that accepts `interface{}`? if there is no other way, ignore this comment.

tranchitella replied:
that's how the validation library works, I cannot do otherwise.

merlin-northern reviewed (commented):
ok I am done with the first round. do you think we should squash everything but the vendors?

tranchitella commented on app/app_iot_hub.go:
```diff
@@ -89,6 +89,43 @@ func (a *app) provisionIoTHubDevice(
+	if err != nil {
+		if htErr, ok := err.(client.HTTPError); ok &&
+			htErr.Code() == http.StatusNotFound {
+			return nil
```
If you are decommissioning a device, and it doesn't exist anymore in Azure IoT Hub, then it's okay.

tranchitella commented on app/app.go:
```diff
@@ -433,17 +420,14 @@ func (a *app) DecommissionDevice(ctx context.Context, deviceID string) error {
-					continue
-				}
-				return errors.Wrap(err, "failed to delete IoT Hub device")
+				return err
```
There is a built-in retry mechanism in workflows, which invokes these end-points.
It will retry up to three times.

tranchitella commented on app/app_iot_core.go:
```diff
@@ -26,22 +26,28 @@ import (
+	}
+	return session.NewSession(&aws.Config{
+		Credentials: credentials.NewStaticCredentials(*accessKeyID, *secretAccessKey, ""),
+	})
```
I disagree, this is part of the contract we sign with the AWS SDK: it either return a session, or an error.
Returning ErrCannotAllocateSession will cause 500, which is exactly the same when we panic.
What's the advantage of returning a custom error? It will only mask the actual issue.

tranchitella commented on app/app_iot_core.go:
```diff
@@ -0,0 +1,108 @@
+		return nil, ErrNoCredentials
+	}
+
+	parts := strings.Split(*endpointURL, ".")
+	region := parts[2] // random-id.iot.us-east-1.amazonaws.com
```
I implemented `getRegionFromEndpoint`

tranchitella commented:
> ok I am done with the first round. do you think we should squash everything but the vendors?

I addressed your comments in the last commit. Yes, we'll squash everything once you're done with the review.

tranchitella commented:
@mender-test-bot start pipeline --fast

tranchitella commented:
@mender-test-bot start pipeline --fast
